cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/sourcecontrol/SourceControlMetadataRefresher.java

```diff
@@ -0,0 +1,106 @@
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.proto.sourcecontrol.RepositoryMeta;
+import io.cdap.cdap.sourcecontrol.operationrunner.SourceControlOperationRunner;
+import io.cdap.cdap.spi.data.StructuredTableContext;
+import io.cdap.cdap.spi.data.TableNotFoundException;
+import io.cdap.cdap.spi.data.transaction.TransactionRunner;
+import io.cdap.cdap.spi.data.transaction.TransactionRunners;
+import io.cdap.cdap.store.RepositoryTable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class manages a mapping of namespaces with repository configurations to their respective
+ * NamespaceSourceControlMetadataRefreshService scheduler instances.
+ */
+public class SourceControlMetadataRefresher {
+  private ConcurrentMap<NamespaceId, Long>
+      lastRefreshTimes = new ConcurrentHashMap<>();
+  private final TransactionRunner transactionRunner;
+  private final SourceControlOperationRunner sourceControlOperationRunner;
+  private final SecureStore secureStore;
+  private static final Logger LOG = LoggerFactory.getLogger(
+      SourceControlMetadataRefresher.class);
+
+  /**
+   * Constructs a SourceControlMetadataRefreshService instance.
+   */
+  @Inject
+  public SourceControlMetadataRefresher(TransactionRunner transactionRunner,
+      SourceControlOperationRunner sourceControlOperationRunner, SecureStore secureStore) {
+    this.transactionRunner = transactionRunner;
+    this.sourceControlOperationRunner = sourceControlOperationRunner;
+    this.secureStore = secureStore;
+  }
+
+  private NamespaceSourceControlMetadataRefresher createRefreshService(NamespaceId namespace) {
+    return new NamespaceSourceControlMetadataRefresher(transactionRunner,
+        sourceControlOperationRunner, namespace);
+  }
+
+  /**
+   * Retrieves the timestamp of the last refresh for the specified namespace.
+   *
+   * @param namespaceId The ID of the namespace for which to retrieve the last refresh time.
+   * @return The timestamp of the last refresh
+   */
+  public long getLastRefreshTime(NamespaceId namespaceId) {
+    return lastRefreshTimes.getOrDefault(namespaceId, 0L);
+  }
+
+  /**
+   * Initiates a manual refresh for the specified namespace. If a refresh service exists for the
+   * namespace, triggers an unforced manual refresh. If no refresh service exists, adds a new
+   * refresh service and starts it. This is triggered when the user calls the list API for namespace
+   * and repository source control metadata.
+   *
+   * @param namespaceId The ID of the namespace for which to run the refresh service.
+   */
+  public void runRefreshService(NamespaceId namespaceId)
+      throws NotFoundException {
+    lastRefreshTimes.putIfAbsent(namespaceId, createRefreshService(namespaceId).refreshMetadata());
+  }
+
+  public RepositoryMeta getRepositoryMeta(NamespaceId namespace)
```

Comment:
This method seems duplicating the one in `SourceControlManagementService`. Why need to duplicate?

Reply:
This class is actually a member of `SourceControlManagementService`. We should move this to a common place. @adrikagupta I think you added a TODO for the same ?